src/v/transform/logging/record_batcher.cc

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "transform/logging/record_batcher.h"
+
+#include "absl/algorithm/container.h"
+#include "model/record.h"
+#include "storage/record_batch_builder.h"
+#include "transform/logging/logger.h"
+#include "utils/vint.h"
+
+namespace transform::logging {
+
+namespace detail {
+class batcher_impl {
+public:
+    batcher_impl() = delete;
+    explicit batcher_impl(size_t batch_max_bytes)
+      : _batch_max_bytes(batch_max_bytes) {}
+    ~batcher_impl() = default;
+    batcher_impl(const batcher_impl&) = delete;
+    batcher_impl& operator=(const batcher_impl&) = delete;
+    batcher_impl(batcher_impl&&) = delete;
+    batcher_impl& operator=(batcher_impl&&) = delete;
+
+    void append(iobuf k, iobuf v) {
+        auto record_size = estimate_record_size(k.size_bytes(), v.size_bytes());
+        if (record_size > max_records_bytes()) {
+            vlog(
+              tlg_log.debug,
+              "Dropping record as Estimated size exceeds configured batch max "
+              "size: {}B vs {}B",
+              record_size,
+              max_records_bytes());
+            return;
+        } else if (record_size >= curr_batch_headroom_bytes()) {
+            roll_batch();
+        }
+        _curr_batch_size += record_size;
+        _builder.add_raw_kv(std::move(k), std::move(v));
+    }
+
+    size_t total_size_bytes() {
+        return absl::c_accumulate(
+          _record_batches, 0, [](size_t acc, const model::record_batch& b) {
+              return acc + b.size_bytes();
+          });
+    }
+
+    ss::chunked_fifo<model::record_batch> finish() {
+        if (!_builder.empty()) {
+            roll_batch();
+        }
+        return std::exchange(_record_batches, {});
+    }
+
+private:
+    inline size_t curr_batch_headroom_bytes() const {
```

Comment:
pretty sure that since c++20 `inline` is implied if it's defined in the class definition like this. https://en.cppreference.com/w/cpp/language/inline#:~:text=A%20function%20defined%20entirely%20inside,(since%20C%2B%2B20).

Reply:
huh, had that store away as only applying to headers, but that doesn't make much sense. good shout, thanks

Reply:
done